`Web docs` - Key value inputs (#2941)
hashicorp/design-system: website/docs/components/form/key-value-inputs/partials/guidelines/guidelines.md

If the maximum number of rows has been reached and the user deletes a row, the alert should be replaced by the add button.

### Deleting rows

If more than one row is present, users can opt to delete any of the rows.

![A user's cursor is hovering over the second row's delete button](/assets/components/form/key-value-inputs/key-value-inputs-ip-delete-any-row.png)

A single row is required at all times. A delete button will only appear for the first row if a second row has been added.

![A single row with a key and value input with no delete button on the side.](/assets/components/form/key-value-inputs/key-value-inputs-single-row.png)

## Footer

The footer supports an add row button and an alert.

### Add row button

The add row button is displayed by default and appends rows to collect additional data and information. The labeling of this button follows our content guidelines of [“add and remove.”](/content/writing-style#add-and-remove)

![A single secondary button with a leading plus icon with the label "Add user"](/assets/components/form/key-value-inputs/key-value-inputs-footer-add-button.png)

> ### Alert

Reviewer: ```suggestion
### Maximum number of rows reached
```

Or something more descriptive than "Alert"

> The alert is used to notify the user of a max row limit being hit. The messaging of this alert follows the language paradigm of “Limit of {max number if word format} {object} has been reached.” 

Reviewer: More directional, positive, and human language.

```suggestion
Use a compact alert to notify the user when they've reached a maximum number of rows. While the content can be customized, we recommend using the following message: "Only {maximum number} {objects} can be added at a time."
```

> This uses the compact neutral alert style so it doesn’t visually compete with any other higher priority validation feedback, like input errors.

Reviewer: ```suggestion
This uses the compact neutral alert, so it doesn’t visually compete with other higher-priority feedback, such as input errors.
```

![A compact neutral alert that says "Limit of fifteen invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-footer-notification.png)

## Responsive behavior

In experiences that are greater than or equal to the [`sm` breakpoint](/foundations/breakpoints), a visible label and helper text are only present in the first row. This is because the form elements are condensed which will help reduce visual clutter.

![Key value input rows with the key and value are side by side along with a delete button to the right. Only one visible label and helper text per grouping of key and value inputs. Add new row button below the group of inputs.](/assets/components/form/key-value-inputs/key-value-inputs-ip-address-desktop.png)

> However, when the viewport is less than the `sm` breakpoint, all field elements stack and occupy 100% of the container and display the label and helper text for their associated inputs. Since the form height increases, the visible labels and helper text orient the user as they are scrolling down and filling in information. If repeating each input’s helper text isn’t useful to the user, consider putting that information in the fieldset helper text instead to reduce the visual noise.

Reviewer: Reading isn't always linear, so it's best to link in both places. 

```suggestion
However, when the viewport is less than the [`sm` breakpoint](/foundations/breakpoints), all field elements stack and occupy 100% of the container and display the label and helper text for their associated inputs. Since the form height increases, the visible labels and helper text orient the user as they are scrolling down and filling in information. If repeating each input’s helper text isn’t useful to the user, consider putting that information in the fieldset helper text instead to reduce the visual noise.
```

## Usage

### When to use

- When two related inputs (often a key and value pair) are repeated for batch submission.
- When collective validation of each associated row is required.

### When not to use

- When inputs in each row are not closely related, use the [form layout helpers](/components/form/layout) instead.

## Header

Key value inputs use a header that includes a legend, helper text, and a generic content slot.

![](/assets/components/form/key-value-inputs/key-value-inputs-legend-content.png)

### Legend

A legend is required to provide context for the overall [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/fieldset). A legend can be configured to render either as its default size, or as a heading depending on where the Key Value Inputs live within the page content.

When placed along with other form components, we recommend using the default legend style provided by the component.

![The key value inputs, which is part of a larger form, has the fieldset legend visible describing to the user what files to upload.](/assets/components/form/key-value-inputs/key-value-inputs-legend-part-of-larger-form.png)

> When using the Key Value Inputs as its own section, we recommend using a heading for the legend. The heading size to use is dependent on the existing heading structure.

Reviewer 2: @majedelass it is incorrect also there :( do you mind open a Jira ticket to revise that guidance (for the RadioCardGroup)? thx

> When using the Key Value Inputs as its own section, we recommend using a heading for the legend. The heading size to use is dependent on the existing heading structure.

Author: Created ticket, will update this doc too.

![Key value inputs are the only form elements on the page, the heading uses display 300 and an H2 tag.](/assets/components/form/key-value-inputs/key-value-inputs-legend-alone-in-form.png)

### Helper text and generic content

The helper text and generic content slot are optional and can be used to provide additional information that applies to all rows.

!!! DO

Use the helper text and generic content slot to reduce the amount of repeated information in each row. This focuses the user’s attention on data submission.

![Helper text used as a means to explain what the toggle in each row does. Generic slot used to include a button with more details.](/assets/components/form/key-value-inputs/key-value-inputs-helper-generic-do.png)

!!!

!!! DONT

Don’t repeat information in each row if it can be explained once at the top of the rows.

![Each row has an input, toggle with helper text enabled and a button in the generic slot. This has repetitive information in each row that over complicates the key value input component's usage.](/assets/components/form/key-value-inputs/key-value-inputs-helper-generic-dont.png)

!!!

> ### Validation

Author: Will move to after the footer. The content was updated, but its position was not.

#### Hitting a row limit

If there is a row limit and the user reaches it, the add button should be replaced with an inline alert notifying them of the limit.

![After the last input row, there is a neutral alert that says "Limit of four invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-row-limit.png)

If the maximum number of rows has been reached and the user deletes a row, the alert should be replaced with the add button.

### Deleting rows

A single row is required at all times. A delete button will only appear for the first row if a second row has been added.

![A single row with a key and value input with no delete button on the side.](/assets/components/form/key-value-inputs/key-value-inputs-single-row.png)

If more than one row is present, users can opt to delete any of the rows.

![A user's cursor is hovering over the second row's delete button](/assets/components/form/key-value-inputs/key-value-inputs-ip-delete-any-row.png)

## Footer

> The footer includes an add button or a compact alert.

Author: From the technical side yes, but from a ux guidelines perspective, we are trying to differentiate the two here.

## Usage

### When to use

- When two related inputs (often a key and value pair) are repeated for batch submission.

> - When collective validation of each associated row is required.

Reviewer 3: @shleewhite What do you think makes most sense related to how the component actually works?

### When not to use

- When inputs in each row are not closely related, use the [form layout helpers](/components/form/layout) instead.

## Header

> Key value inputs use a header that includes a legend, helper text, and a generic content slot.

Reviewer 3: The name of the component should be capitalized for clarity that you are referring to the component throughout the text. (The plural name of the component within sentences reads rather awkwardly, which is made worse when the name isn't capitalized.)

![](/assets/components/form/key-value-inputs/key-value-inputs-legend-content.png)

### Legend

A legend is required to provide context for the overall [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/fieldset).

> Regardless on whether the key value inputs is placed along with other form components, or solo, a legend is always mandetory. 

Reviewer 3: "Regardless of whether the Key Value Inputs component is placed together with other form components, or used on its own, a legend is always mandatory."

#### With other elements

> ![The key value inputs, which is part of a larger form, has the fieldset legend visible describing to the user what files to upload.](/assets/components/form/key-value-inputs/key-value-inputs-legend-part-of-larger-form.png)

Reviewer 3: ```suggestion
![The Key Value Inputs component used as part of a larger form. It includes a Legend explaining to the user what files they should use it to upload.](/assets/components/form/key-value-inputs/key-value-inputs-legend-part-of-larger-form.png)
```

#### Alone in form

> ![Key value inputs are the only form elements on the page.](/assets/components/form/key-value-inputs/key-value-inputs-legend-alone-in-form.png)

Reviewer 3: ```suggestion
![The Key Value Inputs component used as the sole form elements on a page.](/assets/components/form/key-value-inputs/key-value-inputs-legend-alone-in-form.png)
```

### Helper text and generic content

> The helper text and generic content slot are optional and can be used to provide additional information that applies to all rows.

Reviewer 3: ```suggestion
The optional helper text and generic content slots can be used to provide additional information that applies to all rows.
```

!!! DO

Use the helper text and generic content slot to reduce the amount of repeated information in each row. This focuses the user’s attention on data submission.

![Helper text used as a means to explain what the toggle in each row does. Generic slot used to include a button with more details.](/assets/components/form/key-value-inputs/key-value-inputs-helper-generic-do.png)

!!!

!!! DONT

Don’t repeat information in each row if it can be explained once at the top of the rows.

> ![Each row has an input, toggle with helper text enabled and a button in the generic slot. This has repetitive information in each row that over complicates the key value input component's usage.](/assets/components/form/key-value-inputs/key-value-inputs-helper-generic-dont.png)

Reviewer 3: ```suggestion
![Each row has an input, toggle with helper text enabled and a button in the generic slot. This has repetitive information in each row that over complicates the Key Value Input component's usage.](/assets/components/form/key-value-inputs/key-value-inputs-helper-generic-dont.png)
```

!!!

## Input rows

> Each row often has two inputs, one for a key and one for a value. However, there may be exceptions where a single input may be necessary. 

Reviewer 3: Other suggestions...

"Each row typically includes two inputs, ..."
"Each row includes two inputs by default, ..."

Question: Does the Figma component literally include two inputs by default?

![](/assets/components/form/key-value-inputs/key-value-inputs-value-only.png)

### Input types

> Key and value inputs can use various form elements, although there are some common patterns that represent the majority of HashiCorp use cases.

Reviewer 3: ```suggestion
The Key Value Inputs component can contain various types of form elements, although there are some common patterns that represent the majority of HashiCorp use cases.
```

For the key input, this includes:

- Text Input
- Select

For the value input, this includes:

- Text Input
- File Input
- SuperSelect
- Masked Input

### Adding rows

Users can add rows to the set of existing rows via the add button.

![A user clicking the add user button where it shows one row of inputs that turns into two. The single row of inputs doesn't have a delete button.](/assets/components/form/key-value-inputs/key-value-inputs-adding-row.png)

#### Hitting a row limit

If there is a row limit and the user reaches it, the add button should be replaced with an inline alert notifying them of the limit.

![After the last input row, there is a neutral alert that says "Limit of four invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-row-limit.png)

If the maximum number of rows has been reached and the user deletes a row, the alert should be replaced with the add button.

### Delete behavior

!!! Info

> The interactions described in this section will have to be implemented by the consumer.

Reviewer 3: ```suggestion
The interactions described in this section must be implemented by the consumer.
```

!!!

> A single row is required at all times. If that row's content is empty, the delete button will be hidden.

Reviewer 3: ```suggestion
At least one row is must be included. If the row's content is empty, the delete button will be hidden.
```

> ![A single row with a key and value input with no delete button on the side.](/assets/components/form/key-value-inputs/key-value-inputs-single-row-empty-content.png)

Reviewer 3: ```suggestion
![A single row with empty key and value inputs. No delete button is visible.](/assets/components/form/key-value-inputs/key-value-inputs-single-row-empty-content.png)
```

> Once content is filled in either of the two inputs, the delete button will become visible. Clicking it will delete that row's data.

Reviewer 3: ```suggestion
Once content is entered in either of the two inputs, the delete button will become visible. Clicking it will delete the row and its data.
```

> ![A single row with a key and value input with a delete button on the side. The key input has data in it.](/assets/components/form/key-value-inputs/key-value-inputs-single-row-with-content.png)

Reviewer 3: ```suggestion
![A single row with a key and value input. The key input includes entered data. A delete button is visible.](/assets/components/form/key-value-inputs/key-value-inputs-single-row-with-content.png)
```

### Maximum number of rows reached

Use a compact alert to notify the user when they've reached a maximum number of rows. While the content can be customized, we recommend using the following message: "Only {maximum number} {objects} can be added at a time."

This uses the compact neutral alert, so it doesn’t visually compete with other high-priority feedback, such as input errors.

![A compact neutral alert that says "Limit of fifteen invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-footer-notification.png)

### Validation

In addition to form-level validation, if a specific error is associated with the fieldset, it will be displayed below the add button.

![Below the legend and helper text is a compact critical alert showing a fieldset error that the user has reached the maximum number of rows.](/assets/components/form/key-value-inputs/key-value-inputs-fieldset-error.png)

Read more about this in our [form patterns validation documentation](/patterns/form-patterns?tab=validation).

## Responsive behavior

### Small or larger

> In experiences that are greater than or equal to the [medium (`md`) breakpoint](/foundations/breakpoints), the label and helper text are only visible for the first row of inputs. This reduces visual clutter by condensing the form elements.

Reviewer 3: "In experiences that are greater than or equal to the [medium (`md`) breakpoint"

This language sounds awkward to me and "experiences" seems a bit ambiguous. Can you try rewriting using simpler, plainer language?

Author: Will do.

![Key value input rows with the key and value are side by side along with a delete button to the right. Only one visible label and helper text per grouping of key and value inputs. Add new row button below the group of inputs.](/assets/components/form/key-value-inputs/key-value-inputs-ip-address-desktop.png)

### Less than small

> However, when the viewport is less than the [medium (`md`) breakpoint](/foundations/breakpoints), all field elements stack and occupy 100% of the container and display the label and helper text for their associated inputs. As the form height increases, the visible labels and helper text help orient the user as they scroll down and fill in information. If repeating each input’s helper text isn’t useful to the user, consider putting that information in the fieldset helper text instead to reduce the visual noise.

Reviewer 4: [suggestion] Couple of suggestions here, I also removed the "~~~help~~~ orient" because it seemed duplicative with the "helper text", but it does technically "help" them.

```suggestion
When the viewport is less than the [medium (`md`) breakpoint](/foundations/breakpoints), all field elements stack to occupy 100% of the container width and display the label and helper text for their associated inputs. As the form height increases, the visible labels and helper text orient the user as they scroll down and fill in information. If repeating each input’s helper text isn’t useful to the user, consider placing that information in the fieldset helper text instead to reduce the visual noise.
```

### Adding rows

Users can add rows to the set of existing rows via the add button.

![A user clicking the add user button where it shows one row of inputs that turns into two. The single row of inputs doesn't have a delete button.](/assets/components/form/key-value-inputs/key-value-inputs-adding-row.png)

#### Hitting a row limit

If there is a row limit and the user reaches it, the add button should be replaced with an inline alert notifying them of the limit.

![After the last input row, there is a neutral alert that says "Limit of four invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-row-limit.png)

If the maximum number of rows has been reached and the user deletes a row, the alert should be replaced with the add button.

### Delete behavior

The interactions in this section are intended to be implemented by the consumer. At least one row should remain visible at all times.

#### Single row behavior

If the row's inputs are empty, the delete button will be hidden.

![A single row with empty key and value inputs. No delete button is visible.](/assets/components/form/key-value-inputs/key-value-inputs-single-row-empty-content.png)

> Once content is entered in either of the two inputs, the delete button will become visible. Clicking it will remove the row's data and then hide the button.

Reviewer 5: [nit]
```suggestion
Once content is entered in either of the two inputs, the delete button will become visible. Clicking it will remove the row's data and hide the button.
```

### Input types

The Key Value Inputs component can contain various types of form elements, although there are some common patterns that represent the majority of HashiCorp use cases.

For the key input, this includes:

- Text Input
- Select

For the value input, this includes:

- Text Input
- File Input
- SuperSelect
- Masked Input

### Adding rows

Users can add rows to the set of existing rows via the add button.

![A user clicking the add user button where it shows one row of inputs that turns into two. The single row of inputs doesn't have a delete button.](/assets/components/form/key-value-inputs/key-value-inputs-adding-row.png)

#### Hitting a row limit

If there is a row limit and the user reaches it, the add button should be replaced with an inline alert notifying them of the limit.

![After the last input row, there is a neutral alert that says "Limit of four invites has been reached."](/assets/components/form/key-value-inputs/key-value-inputs-row-limit.png)

If the maximum number of rows has been reached and the user deletes a row, the alert should be replaced with the add button.

### Delete behavior

> !!! Info
> 
> The interactions described in this section must be implemented by the consumer.
> 
> !!!
> 
> At least one row is required at all times.

Reviewer 6: Since this sentence states a consumer responsibility and does not warn about an error state or prompt immediate action, I think it would be better as inline text. It provides context but doesn’t carry enough urgency or importance to justify a banner. 

```suggestion
The interactions in this section are intended to be implemented by the consumer. At least one row should remain visible at all times.
```

Author: Thanks for the catch, updating.
